### ISSUE TITLE
c++ SGD and python SGD don't match

### ISSUE THREAD

**Issue by an-kumar** [pytorch/pytorch#15530]
## 🐛 Bug

I would consider this a bug as the stated goals of the c++ frontend is to track the python library.

The difference is itself stated in the comments in the python library:

"
.. note::
        The implementation of SGD with Momentum/Nesterov subtly differs from
        Sutskever et. al. and implementations in some other frameworks.
        Considering the specific case of Momentum, the update can be written as
        .. math::
                  v = \rho * v + g \\
                  p = p - lr * v
        where p, g, v and :math:`\rho` denote the parameters, gradient,
        velocity, and momentum respectively.
        This is in contrast to Sutskever et. al. and
        other frameworks which employ an update of the form
        .. math::
             v = \rho * v + lr * g \\
             p = p - v
        The Nesterov version is analogously modified.
"

The C++ API implements SGD in the way that Sutskever et al and other frameworks do, which is to only apply the learning rate to the gradient, not to the entire step.

I think the discrepancy should be fixed.

cc @goldsborough

**Comment by goldsborough** [comment]
Thanks for reporting this. I'm slightly surprised the tests don't catch this because we actually compare the C++ optimizers against the Python optimizers with quite a lot of effort (comparing parameter values at every optimization step). I think the first step to solving this would be to make the tests fail.

@an-kumar do you want to work on this? Otherwise we can take care of it. We should fix this ASAP either way.

**Comment by goldsborough** [comment]
Yes the instructions for building from source are in https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md

**Comment by an-kumar** [comment]
So, here's what I'm thinking in order to do this correctly. It's a slightly larger PR than what I originally had in mind so I wanted to make sure we're on the same page.

As mentioned the discrepancy only occurs when the LR changes. To fix and test I therefore need to change the learning rate. Python has the `lr_scheduler` abstraction that does this, which I would use to set up the parameters for the exact match tests.

I want to just implement `lr_scheduler` in c++ as well. The python `lr_scheduler` works by mutating the optimizer's 'param_group' member variable. There are a few options on how to bring this to c++:

1- templatize LRScheduler on the optimizer type and mutate `optimizer.options`. every optimizer has `options.learning_rate_` which would be what is mutated. The downside of this is lack of clarity, where we're just assuming that optimizer classes need to have `options.learning_rate_` be valid

2- implement a virtual function on the `Optimizer` abstract class called `setLearningRate()`. The downside of this is theoretically we may want to allow mutation of other parameters other than learning rate, but those parameters would be derived-class-specific, so wouldnt be appropriate to put in the abstract class.

both of the above options would work fine when / if we introduce parameter groups in c++.

I think 2 is better, every optimizer has a learning rate anyways so it seems fine to just put it in the abstract class.

what do you think?

**Comment by goldsborough** [comment]
While I would generally welcome a LR scheduler concept for the C++ frontend, I have two thoughts:
- Implementing a whole LR scheduler feature just to write a test seems a bit overkill. Just changing `options.learning_rate_` would be fine for the test I think?
- I'm not sure parameter groups are necessary -- AFAIK they simply allow using different optimization options for different groups of parameters in a single model. In the C++ optimizers we don't have a concept of parameter groups (i.e. there is only one group, which is all parameters). But this concept should be orthogonal to LR scheduling.

**Comment by goldsborough** [comment]
I think it'd be awesome to have a LR scheduler in the C++ frontend and if you'd like to design and implement it, that would be greatly appreciate it and I'll review it for you.

I do think the SGD fix can be separate. You can just re-implement the Python algorithm actually, I did that a few days ago when investigating the NaN issues (which are now fixed by the way).

**Comment by an-kumar** [comment]
I made the sgd fix as a separate pr. I didn't do the learning rate changes -- I'm just depending on the old SGD tests to confirm that the new implementation still matches. https://github.com/pytorch/pytorch/pull/15840

I'll work on the LR scheduler as a separate PR and add extra tests to confirm that even when the LR changes the various optimizers match python.
